### PR TITLE
Submission settings columns

### PR DESCRIPTION
**Pull request by hyena**
These changes affect a lot of logic so it needs a careful once over. I've tested some of these but it could benefit from other people pocking too:

Make sure that you can't view hidden submissions (checked)
Can't view friends-only submissions without being friends (checked briefly)
Critique shows up correctly (checked)
Friends-only shows up in read (checked)
Moderation tools work (checked somewhat)

### CODE REVIEW

File: weasyl/moderation.py [hunk 1]
```diff
@@ -409,8 +409,11 @@ def submissionsbyuser(userid, form):
     if userid not in staff.MODS:
         raise WeasylError("Unexpected")
 
+    # TODO(hyena): su.settings has been removed from here. Replace it with the individual columns and fix up
```
Comment (hyena) [inline weasyl/moderation.py]: Note to self: You've done this; remove the comment.

File: weasyl/errorcode.py [hunk 2]
```diff
@@ -149,6 +151,7 @@
         "The link you followed does not appear to be valid. You may have already added these premium terms "
         "to your account, or you may have copied the link incorrectly."),
     "tooManyPreferenceTags": "You cannot have more than 50 preference tags.",
+    # TODO(hyena): The pointer on this could be more explicit. HOW should the user report the bug? To WHOM?
```
Comment (charmander) [inline weasyl/errorcode.py]: This comment appears to go with `logicalError`.

File: weasyl/moderation.py [hunk 3]
```diff
@@ -631,6 +644,11 @@ def action(tbl):
                 tbl.update()
                 .values(settings=tbl.c.settings.op('||')('h'))
                 .where(tbl.c.settings.op('!~')('h')))
+        def split_columns_action(tbl):
+            return (
+                tbl.update()
+                .values(hidden=True)
+                .where(tbl.c.hidden.isnot(True)))
```
Comment (charmander) [inline weasyl/moderation.py]: Does this work as `.where(~tbl.c.hidden)`?

File: weasyl/favorite.py [hunk 4]
```diff
@@ -190,14 +190,18 @@ def insert(userid, submitid=None, charid=None, journalid=None):
...
+                          [content_table, id_field, target], options="single")
+    else:
+        query = d.execute("SELECT userid, settings FROM %s WHERE %s = %i",
+                          [content_table, id_field, target], options="single")
 
     if not query:
         raise WeasylError("TargetRecordMissing")
     elif userid == query[0]:
         raise WeasylError("CannotSelfFavorite")
-    elif "f" in query[1] and not frienduser.check(userid, query[0]):
```
Comment (charmander) [inline weasyl/favorite.py:200]: Not introduced in this PR, but a reminder for after: there should probably be a check for `hidden`.

File: weasyl/favorite.py [hunk 5]
```diff
@@ -190,14 +190,18 @@ def insert(userid, submitid=None, charid=None, journalid=None):
     else:
         content_table, id_field, target = "journal", "journalid", journalid
 
-    query = d.execute("SELECT userid, settings FROM %s WHERE %s = %i",
-                      [content_table, id_field, target], options="single")
+    if submitid:
+        query = d.execute("SELECT userid, friends_only FROM %s WHERE %s = %i",
```
Comment (charmander) [inline weasyl/favorite.py:194]: Branch can be simplified.

File: weasyl/favorite.py [hunk 6]
```diff
@@ -190,14 +190,18 @@ def insert(userid, submitid=None, charid=None, journalid=None):
...
+    else:
+        query = d.execute("SELECT userid, settings FROM %s WHERE %s = %i",
+                          [content_table, id_field, target], options="single")
 
     if not query:
         raise WeasylError("TargetRecordMissing")
     elif userid == query[0]:
         raise WeasylError("CannotSelfFavorite")
-    elif "f" in query[1] and not frienduser.check(userid, query[0]):
+    elif (submitid and query[1] or not submitid and "f" in query[1]) and not frienduser.check(userid, query[0]):
```
Comment (charmander) [inline weasyl/favorite.py:204]: Selecting `settings ~ 'f'` would avoid this.

File: libweasyl/libweasyl/alembic/versions/30ddd5fc6d26_submission_settings_columns.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        'submission',
+        sa.Column('friends_only', sa.Boolean(), nullable=False, server_default='f'),
+    )
+    op.add_column(
+        'submission',
+        sa.Column('critique', sa.Boolean(), nullable=False, server_default='f'),
+    )
+    op.add_column(
+        'submission',
+        sa.Column('embed_type', embed_types, nullable='f'),
```
Comment (charmander) [inline libweasyl/libweasyl/alembic/versions/30ddd5fc6d26_submission_settings_columns.py:39]: `nullable='f'`?

**Review by charmander — CHANGES_REQUESTED** [review]
To check:

- `is_hidden` in weasyl/templates/modcontrol/reports.html

File: weasyl/submission.py [hunk 8]
```diff
@@ -535,9 +519,12 @@ def is_hidden(submitid):
```
[the remark this answers is not in the document]
Comment (charmander) [reply weasyl/submission.py:537]: (no longer exists)

**Review by charmander — APPROVED** [review]

**Pull request merged**